lib/livebook/runtime/evaluator/io_proxy.ex

```diff
@@ -128,6 +152,19 @@ defmodule Livebook.Runtime.Evaluator.IOProxy do
     {:noreply, flush_buffer(state)}
   end
 
+  def handle_info({:DOWN, ref, :process, _pid, reason}, state)
+      when ref == state.evaluator_monitor do
+    cleanup(state)
+    {:stop, reason, state}
+  end
+
+  defp cleanup(state) do
+    # Remove all modules defined during evaluation
+    for module <- state.modules_defined, Code.ensure_loaded?(module) do
```

Comment:
@josevalim since we actually gather tracer data in the group leader, we just accumulate a set with all modules we ever defined, and then filter out those that are no longer there. We expect this set to have just a couple elements, so that should be fine?

Reply:
Code.ensure_loaded? may attempt to load the module. `function_exported?(module, :module_info, 1)` should be a faster check.